File: framework/languages/dotnet/e2e/YieldDispatch/Client/Scenarios/YdB2SameActorReentryScenario.cs
using YieldDispatch.Client.Support;
using YieldDispatch.Shared;

namespace YieldDispatch.Client.Scenarios;

internal static class YdB2SameActorReentryScenario
{
    public static async Task RunAsync(
        IZlinkStreamConnector client,
        YieldActorScenarioContext actors)
    {
        var requestId = $"YD-B2-{Guid.NewGuid():N}";
        var yield = client.Request(new ActorYieldReq(requestId, 350))
            .PacketName("ActorYieldReq")
            .Metadata(YieldDispatchNames.ActorIdMetadata, actors.ActorA)
            .Timeout(TimeSpan.FromSeconds(30))
            .Async<ActorYieldReply>();
        await Task.Delay(75);
        var fast = client.Request(new ActorFastReq(requestId, "b2-fast"))
            .PacketName("ActorFastReq")
            .Metadata(YieldDispatchNames.ActorIdMetadata, actors.ActorA)
            .Timeout(TimeSpan.FromSeconds(30))
            .Async<ActorYieldReply>();
        await Task.WhenAll(yield.AsTask(), fast.AsTask());

        var evidence = await client.Request(new YieldEvidenceReq(requestId))
            .PacketName("YieldEvidenceReq")
            .Metadata(YieldDispatchNames.TargetNodeRidMetadata, "play-a")
            .Timeout(TimeSpan.FromSeconds(30))
            .Async<YieldEvidenceReply>();
        ScenarioAssert.ContainsExactRequestInOrder(evidence.Evidence, requestId, [
            "actor-yield-started",
            "actor-yield-released",
            "actor-yield-resumed",
            "actor-yield-completed",
            "actor-fast-started",
            "actor-fast-completed"
        ]);
    }
}

File: framework/languages/node/e2e/YieldDispatch/Client/Scenarios/yd-b2-same-actor-reentry-scenario.ts
import type {
  ActorFastMsg,
  ActorYieldReq,
  YieldEvidenceRes,
  YieldEvidenceReq
} from '../../Shared/messages';
import { YieldDispatchNames } from '../../Shared/messages';
import { containsRequestMarkersInOrder } from '../Support/scenario-assert';
import type { ZlinkStreamConnector } from '@zlink-systems/stream-connector';
import type { YieldActorScenarioContext } from './yd-b1-other-actor-progress-scenario';

export async function runYdB2(
  client: ZlinkStreamConnector,
  actors: YieldActorScenarioContext
): Promise<string> {
  const requestId = `YD-B2-${uniqueId()}`;
  const yielding = client
    .request({ requestId, delayMs: 350 } satisfies ActorYieldReq)
    .packetName('ActorYieldReq')
    .metadata(YieldDispatchNames.actorIdMetadata, actors.actorA)
    .timeout(30000)
    .submit();
  await new Promise((resolve) => setTimeout(resolve, 75));
  await client
    .send({ requestId, marker: 'b2-fast' } satisfies ActorFastMsg)
    .packetName('ActorFastMsg')
    .metadata(YieldDispatchNames.actorIdMetadata, actors.actorA)
    .submit();
  await yielding;

  const evidence = await client
    .request({ requestId } satisfies YieldEvidenceReq)
    .packetName('YieldEvidenceReq')
    .metadata(YieldDispatchNames.targetNodeRidMetadata, 'play-a')
    .timeout(30000)
    .submit<YieldEvidenceRes>();
  containsRequestMarkersInOrder(evidence.evidence, requestId, [
    'actor-yield-started',
    'actor-yield-released',
    'actor-yield-resumed',
    'actor-yield-completed',
    'actor-fast-started',
    'actor-fast-completed'
  ], 'YD-B2 marker order mismatch.');
  console.log('scenario YD-B2 passed');
  return requestId;
}

function uniqueId(): string {
  return `${Date.now().toString(16)}-${Math.random().toString(16).slice(2)}`;
}
